src/ray/raylet/node_manager.cc

```diff
@@ -3013,6 +3022,61 @@ void NodeManager::FinishAssignTask(const std::shared_ptr<Worker> &worker,
   }
 }
 
+void NodeManager::ProcessSubscribePlasma(
+    const std::shared_ptr<LocalClientConnection> &client, const uint8_t *message_data) {
+  std::shared_ptr<Worker> associated_worker = worker_pool_.GetRegisteredWorker(client);
+  if (associated_worker == nullptr) {
+    associated_worker = worker_pool_.GetRegisteredDriver(client);
+  }
+  if (associated_worker == nullptr) {
+    RAY_LOG(ERROR) << "No worker exists for CoreWorker with client: "
+                   << client->DebugString();
+    return;
+  }
+
+  auto message = flatbuffers::GetRoot<protocol::SubscribePlasma>(message_data);
+  ObjectID id = from_flatbuf<ObjectID>(*message->object_id());
+  {
+    absl::MutexLock guard(&plasma_object_lock_);
+    if (!async_plasma_objects_.contains(id)) {
+      async_plasma_objects_.emplace(id, absl::flat_hash_set<std::shared_ptr<Worker>>());
+    }
+    if (!async_plasma_objects_[id].contains(associated_worker)) {
+      async_plasma_objects_[id].insert(associated_worker);
+    }
+  }
+}
+
+ray::Status NodeManager::SetupPlasmaSubscription() {
+  return object_manager_.SubscribeObjAdded(
+      [this](const object_manager::protocol::ObjectInfoT &object_info) {
+        ObjectID object_id = ObjectID::FromPlasmaIdBinary(object_info.object_id);
+        auto waiting_workers = absl::flat_hash_set<std::shared_ptr<Worker>>();
+        {
+          absl::MutexLock guard(&plasma_object_lock_);
+          auto waiting = this->async_plasma_objects_.extract(object_id);
+          if (!waiting.empty()) {
+            waiting_workers.swap(waiting.mapped());
+          }
```

Comment:
this doesn't pop the entry in `async_plasma_objects_` map right? just set it to empty set? make sure to pop the corresponding entry because we don't want the map grow infinitely large.